Improve the partial derivatives of tfp_math.betainc

leandrolcampos opened the pull request:
This PR tries to improve the partial derivatives of ` tfp_math.betainc(a, b, x)` with respect to `a` and `b`.

Some of the improvement opportunities I identified in the current implementation of these partial derivatives are:
- When dtype is `float64`:
  - Moderate loss of accuracy when `a` and `b` are uniformly distributed in the interval (0, 10).
  - When `a` and `b` are very small, gradients contain a very high proportion of NaN.
- When dtype is `float32`:
  - Heavy loss of accuracy in all domains evaluated.
  - Gradients contain NaN (sometimes infinite as well) in all domains evaluated.
  - When `a` and `b` are very small, gradients contain only NaN.

Please see the following notebook:
https://gist.github.com/leandrolcampos/c5a77b0f24ee8d2a38c9ea2f8ddc3c77

The alternative implementation I propose addresses all these improvement opportunities. Please see the following notebook:
https://gist.github.com/leandrolcampos/776048810f9c4ae18fc9598f54543057

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+  small = tf.sqrt(tiny)
+
+  def continued_fraction_evaluation(should_stop, iteration, values, gradients):
+    c, d, f = values
```
Can these be renamed to be more descriptive?

leandrolcampos replied:
Sorry for the non-suggestive names. Here I follow the notation used on the Wikipedia page about the [Lentz algorithm](https://en.wikipedia.org/wiki/Lentz%27s_algorithm). On this page, `f_n` is the convergent `f_n = A_n / B_n = f_{n-1} * C_n * D_n` and the quantities `C_n` and `D_n` are the ratios `C_n = A_n / A_{n-1}` and `D_n = B_{n-1} / B_n`. The partial numerator `A_n` and denominator `B_n` are given by the [fundamental recurrence formulas](https://en.wikipedia.org/wiki/Generalized_continued_fraction#Fundamental_recurrence_formulas). Do you have any suggestions?

Maybe:
`convergent` for `f`
`ratio_numerators` for `c`
`ratio_denominators` for `d`
`dconvergent` for `f_grad`
`dratio_numerators` for `c_grad`
`dratio_denominators` for `d_grad`

srvasude replied:
Those names sound good to me! I think having a comment which specifies that map (in case someone is familiar with Lentz's algorithm) would also be welcome (so using the more descriptive names, but saying that ratio_numerators = C_n in a comment.

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+  cf, cf_grad_a, cf_grad_b = _betainc_modified_lentz_method(
+      a, b, x, dtype, where)
+
+  more_terms = tf.math.exp(
```
Rename? Maybe normalization? Here and elsewhere.

leandrolcampos replied:
Done.

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+    new_product = product * factor
+    term = new_product / apn
+    new_dpdb = factor * dpdb - product * x_div_n
+    new_da = da - new_product / (apn * apn)
```
Use `tf.math.square`

leandrolcampos replied:
Done (here and elsewhere).

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+  safe_x = tf.where(where, x, half)
+
+  # When the condition C1 is false, we apply the symmetry relation given
+  # here: http://dlmf.nist.gov/8.17.E4
```
It would be nice (here and elsewhere) to inline the relationship, that way it is easier to see at a glance what the variable / code is referring to.

leandrolcampos replied:
I'm not sure if I understood this comment. Is it about making the symmetry relationship explicit in the code comment, instead of using a link?

srvasude replied:
Sorry, yes I mean making it explicit in the code comment (and leaving the link in case someone wants to read more).

leandrolcampos replied:
Done.

srvasude reviewed (commented):
Hi, thanks for this awesome PR! Overall looks good to me, and most of my comments have to do with naming (basically preferring more verbose names over short names so as to keep the code more maintainable).

leandrolcampos commented on tensorflow_probability/python/math/special.py:
```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+    return value, gradient
+
+  # Assume iteration is always greater than 1.
+  return tf.cond(
```
If we use `tf.where` instead of `tf.cond`, I understand that the tuple `(values, gradients)` will be computed twice for each call of `_betainc_nth_partial_numerator`, regardless of whether `iteration` is odd or even. Is my understanding correct? If so, why should we prefer `tf.where` over `tf.cond`?

srvasude replied:
In general we prefer using `tf.where`, especially in cases like this that are relatively cheap. `tf.cond` creates a runtime branch to decide which expression to compute. This can be expensive, and often `tf.where` (which is branchless) will be more performant on accelators like GPU or Cloud TPU.

leandrolcampos replied:
After refactoring the continued fraction evaluation procedures, `tf.cond` is no longer needed.

leandrolcampos commented:
Hi, @srvasude! Many thanks for your time reviewing this PR. I have a few questions about your comments above. Can you take a look at them?

leandrolcampos commented:
Hi, @srvasude! I addressed all of your comments. Can you please review the PR again?

After refactoring the continued fraction evaluation procedures, `tf.cond` is no longer needed.

Thank you very much for your guidance!

leandrolcampos commented on tensorflow_probability/python/math/special_test.py:
```diff
@@ -194,27 +196,300 @@ def testBetaincBroadcast(self):
+    strm = test_util.test_seed_stream()
+    a = tfp.distributions.Uniform(
+        high=np.float32(5.)).sample(100, strm())
+    a = self.evaluate(a)
```
Done.

leandrolcampos commented on tensorflow_probability/python/math/special.py:
```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+
+  # The partial derivatives of betainc with respect to a and b are computed
+  # using forward mode.
+  region_ps = ((x < a / (a + b)) & (b * x <= 1.) & (x <= 0.95) |
```
Done. I replaced `region_ps` by `use_power_series`.

leandrolcampos commented:
Hi, @srvasude! If I still have a chance, I'd like to ask you two more questions.

1. I noticed that I should have used 4 spaces for indentation on lines 218 and 222. If I fix this now, GitHub will run all the checks just for that. Can you fix this during the internal review? Or do you prefer that I do it myself?

2. Following the current implementation, I call `tf.convert_to_tensor` on `_betainc_naive`, `_betainc_partials`, and on `betainc`. Is this necessary? Or is it enough to just call `tf.convert_to_tensor` in `betainc`, immediately after name_scope?

All the best,

srvasude reviewed (approved):
Thanks so much for these changes, and the thought and care in your analysis!

I'll be merging these in (and looking forward to any future work on this and related functions :) ).
